Fix for initialization and finalization sections

lorenzogentile404 opened the pull request:
- spec: https://github.com/Consensys/linea-specification/issues/11
- constraints: https://github.com/Consensys/linea-constraints/pull/517.

lorenzogentile404 committed:
update constraints

lorenzogentile404 committed:
update hub trace file and fix corresponding tracing

lorenzogentile404 committed:
update initialization section

lorenzogentile404 committed:
manage reverted transaction case for initialization section

lorenzogentile404 committed:
update finalization section

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -127,20 +148,52 @@ public TxInitializationSection(Hub hub, WorldView world) {
     this.addFragments(
         ImcFragment.forTxInit(hub), ContextFragment.initializeExecutionContext(hub), txFragment);
 
+    boolean isRevertedTransaction = false; // TODO: how to get this?
```
You have to defer to post transaction to finish the initialization section.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -37,7 +37,9 @@
-  @Getter private final AccountSnapshot senderAfterPayingForTransaction;
+  @Getter private final AccountSnapshot senderAfterPayingForGas;
+  @Getter private final AccountSnapshot senderAfterPayingForValue;
+  @Getter private final AccountSnapshot senderAfterPayingForGasAndValue;
```
I would suggest, for the sheer purpose of readability, to have the following snapshots be part of the class definition
```java
private final AccountSnapshot senderGasPayment
private final AccountSnapshot senderGasPaymentNew
private final AccountSnapshot senderValueTransfer
private final AccountSnapshot senderValueTransferNew
private final AccountSnapshot recipientValueReception
private final AccountSnapshot recipientValueReceptionNew
private AccountSnapshot senderUndoingValueTransfer
private AccountSnapshot senderUndoingValueTransferNew
private AccountSnapshot recipientUndoingValueReception
private AccountSnapshot recipientUndoingValueReceptionNew
```
I've started explicitly introducing all required `AccountSnapshot`'s as it became unmanageable otherwise in other sections (`CallSection` or `CreateSection` for instance.) As we handle a couple of hundred transactions total at a time max this isn't too much of an imposition on memory etc. And it makes intent more clear.

lorenzogentile404 replied:
Ok.

lorenzogentile404 replied:
Do we actually need all of them? `senderGasPaymentNew` and `senderValueTransfer` should be the same. Also, I find clearer the notation "before" and "after". Anyway, let's discuss it tomorrow.

lorenzogentile404 replied:
Also, `senderValueTransferNew` and `senderUndoingValueTransfer`, `recipientValueReceptionNew` and `recipientUndoingValueReception` should be the same.

lorenzogentile404 replied:
Note: whenever variables represent the same thing, create a deep copy of the underlying objects.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -37,7 +37,9 @@
 
 public class TxInitializationSection extends TraceSection {
```
This has to `implements PostTransactionDefer` or whatever it's called.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxFinalizationSection.java:
For this it would be ideal if we could get access to the refunded gas from Besu directly. I'm not sure we have it available here, or how, but we would need the values that are available here https://github.com/hyperledger/besu/blob/98780efd15ee3223de2c0cf2d144c8aaee30d5db/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetTransactionProcessor.java#L502

lu-pinto replied:
The outermost frame should give you the refunds I believe, would need to double check it. It also gives you the selfDestructs and you can infer the maths as in the class you posted there.
I don't think you care about code delegation since you are on the London fork

lorenzogentile404 committed:
improve naming and structure

lorenzogentile404 committed:
add resolvePostTransaction

lorenzogentile404 committed:
minor changes

lorenzogentile404 committed:
Merge branch 'arith-dev' into 1648-initialization-and-finalization-sections-fix

lorenzogentile404 committed:
update constraints

lorenzogentile404 committed:
add scheduleForPostTransaction

lorenzogentile404 committed:
remove senderAfterPayingForValue

lorenzogentile404 committed:
extract gas cost

lorenzogentile404 committed:
complete initialization section

lorenzogentile404 committed:
add comment

lorenzogentile404 committed:
improve naming in finalization section

lorenzogentile404 committed:
Merge branch 'arith-dev' into 1648-initialization-and-finalization-sections-fix

lorenzogentile404 committed:
update constraints

lorenzogentile404 committed:
add comments

OlivierBBB committed:
fix: record HUB_STAMP for future undoing

OlivierBBB committed:
ras

OlivierBBB committed:
fix: the final context fragment has to be created at the start

OlivierBBB committed:
fix: DOM/SUB stamps of the revert have to use the logged hubStamp

OlivierBBB committed:
ras: made isSelfCredit(Hub) a method

OlivierBBB committed:
fix: the balance increment was off

OlivierBBB committed:
fix: 3 DOM/SUB stamps are required initially, not 2

OlivierBBB committed:
ras

OlivierBBB committed:
ras

OlivierBBB committed:
fix: more on initialization

OlivierBBB committed:
ras: wtarted working on finalization

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -38,73 +38,66 @@
             deploymentInfo.getDeploymentStatus(senderAddress));
-    senderDomSubStamps = DomSubStampsSubFragment.standardDomSubStamps(hub.stamp(), 0);
+    senderGasPaymentNew =
+        senderGasPayment.deepCopy().decrementBalanceBy(gasCost).turnOnWarmth().raiseNonceByOne();
```
Raising of the nonce.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -38,73 +38,66 @@
-    senderValueTransferNew =
-        senderValueTransfer
-            .deepCopy()
-            .decrementBalanceBy(valueAndGasCost)
```
Gas was already paid for.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -142,80 +129,82 @@ public TxInitializationSection(Hub hub, WorldView world) {
       recipientValueReceptionNew
           .raiseNonceByOne()
-          .incrementBalanceBy(incrementToApplyToRecipientBalance)
+          .incrementBalanceBy(value)
```
You always provide `value` to the recipient.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -38,73 +38,66 @@
+    super(hub, (short) 8);
     hub.defers().scheduleForPostTransaction(this);
 
+    hubStamp = hub.stamp();
```
Required for the `DOM`/`SUB` stamps in case of a `REVERT`ed transaction. The `HUB_STAMP` that you use is the one of the start of the transaction (in the `SUB_STAMP`.)

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -142,80 +129,82 @@ public TxInitializationSection(Hub hub, WorldView world) {
-            senderValueTransfer, senderValueTransferNew, senderDomSubStamps));
+            senderValueTransfer,
+            senderValueTransferNew,
+            DomSubStampsSubFragment.standardDomSubStamps(hubStamp, 1)));
```
We need 2 different time stamps for the two actions (paying for gas / transferring value)

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -142,80 +129,82 @@ public TxInitializationSection(Hub hub, WorldView world) {
             .requiresRomlex(true));
 
+    initializationContextFragment = ContextFragment.initializeExecutionContext(hub);
+
```
We create the ContextFragment that will be inserted at the very end of the transaction processing.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -142,80 +129,82 @@ public TxInitializationSection(Hub hub, WorldView world) {
-          senderUndoingValueTransfer.deepCopy().setDeploymentNumber(hub).incrementBalanceBy(value);
 
+      senderUndoingValueTransfer = senderValueTransferNew.deepCopy().setDeploymentNumber(hub);
+      senderUndoingValueTransferNew = senderValueTransfer.deepCopy().setDeploymentNumber(hub);
```
Doing things (_de_crementing by `value`) in reverse order amounts to doing the opposite (_in_crementing by `value`)

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/TxInitializationSection.java:
```diff
@@ -142,80 +129,82 @@ public TxInitializationSection(Hub hub, WorldView world) {
               recipientUndoingValueReception,
               recipientUndoingValueReceptionNew,
-              recipientDomSubStamps));
+              DomSubStampsSubFragment.revertWithCurrentDomSubStamps(hubStamp, revertStamp, 4)));
```
Here we use the `hub.stamp()` snapshot from transaction initialization.

OlivierBBB committed:
ras: scheduleForPostTransaction => scheduleForEndTransation

OlivierBBB committed:
ras: isSelfCredit => senderIsRecipient

OlivierBBB committed:
fix: TX_FINL implementation

OlivierBBB committed:
ras: names + other

OlivierBBB committed:
ras: switch to master for constraints in order to merge arith-dev

OlivierBBB committed:
Merge branch 'arith-dev' into 1648-initialization-and-finalization-sections-fix

OlivierBBB committed:
ras: regenerate the Trace.java files for HUB and BLOCKDATA

OlivierBBB committed:
fix: the TX_INIT section requires to scheduleForContextEntry its MISC fragment

The purpose is: log the CCSR and CCRS of the child context (the root
context of the transaction)

OlivierBBB committed:
fix: type change in the HUB

OlivierBBB committed:
fix: schedule the TX_INIT's MISC fragment for ContextEntry and allowing the HUB to resolve these

OlivierBBB committed:
fix: rebuilt trace files

OlivierBBB committed:
fix: using ETHEREUM_GAS_LIMIT_XXX constants

OlivierBBB committed:
fix: constraints update

OlivierBBB committed:
ras: constraints commit update

OlivierBBB committed:
feat: more ignored tests

OlivierBBB committed:
fix: ignore MODEXP tests where bbs, ebs or mbs is > 512

OlivierBBB committed:
Merge branch 'arith-dev' into 1648-initialization-and-finalization-sections-fix

OlivierBBB committed:
fix: erroneous traceEndConflation in CorsetBlockProcessor

OlivierBBB committed:
fix: raise jupiter timeout to 15 minutes

OlivierBBB committed:
ras: direct method for CFI extraction

OlivierBBB committed:
ras: remove redundant `setWarmthTo(true)`

This is redundant as the warmth of the coinbase in the

  senderIsCoinbase

case is already checkState'd just above.

OlivierBBB committed:
fix: more ignored tests

- re-orgs
- forks
- "new chain"
- side chain

OlivierBBB committed:
ras

OlivierBBB committed:
ras: simplified refund computation

OlivierBBB committed:
fix: constraints SSTORE fix + ras

lorenzogentile404 committed:
update constraints

lorenzogentile404 committed:
Merge branch '1648-initialization-and-finalization-sections-fix' of github.com:Consensys/linea-tracer into 1648-initialization-and-finalization-sections-fix

OlivierBBB committed:
fix: more ignored tests (deployment tx leads to collision)

OlivierBBB committed:
ras: minor changes in TX_INIT

OlivierBBB committed:
feat: ignoring a test triggering EIP-7610

Also: added error message when this or similar issues is triggered

OlivierBBB committed:
spotless

OlivierBBB committed:
feat: ignoring a test (see #1678)

OlivierBBB committed:
ras

OlivierBBB committed:
fix: the "1th value 256 does not fit a byte" issue

OlivierBBB committed:
feat: ignoring the last tests blowing up the BLOCKDATA module

OlivierBBB committed:
ras: basic print statement for MODEXP calls with illegal arguments

OlivierBBB committed:
ras: extracted test from Reference tests

OlivierBBB committed:
feat: more test vectors that we ignore

These are related to invalid MODEXP arguments

OlivierBBB commented:
For ignored tests of commit https://github.com/Consensys/linea-tracer/pull/1657/commits/76bbf00a76c44aa891843d2583a764a3ceee995c see https://github.com/Consensys/linea-tracer/issues/1578

OlivierBBB committed:
fix: updating constraints commit

OlivierBBB committed:
ras: switch from ETHEREUM_GAS_LIMI_XXX back to LINEA_GAS_LIMIT_XXX constants

OlivierBBB committed:
spotless

OlivierBBB committed:
fix: new boolean "coinbaseWarmthAtTransactionEnd" field in HUB

This boolean records the warmth of the coinbase address at transaction
end. The old way of doing things didn't work with

  BlockchainReferenceTest_331/randomStatetest643

where sender == coinbase. And likely this would have blown up in general
with consistency constraints on.

OlivierBBB committed:
fix: safe addition for upper bound of GAS_LIMIT computation in BLOCKDATA

OlivierBBB committed:
fix: in BLOCKDATA maxDeviation must be made unsigned

OlivierBBB committed:
ras

OlivierBBB committed:
fix: reworking of the TxSkipSection

We now extract the coinbase address from a MessageFrame. This ought to
fix an issue with replay tests with a TX_SKIP type of transaction.
Indeed, for some reason (maybe related to using clique and this changing
the "miner" field of the block header) running

  block.getCoinbaseAddress()

or so returned 0x00 rather than the expected 0x8f8.

Also greatly simplified this TraceSection.

OlivierBBB committed:
fix(disgusting): coinbaseAddress in HUB

OlivierBBB committed:
fix: use the Hub#coinbaseAddress in BLOCKDATA

OlivierBBB committed:
fix: use the Hub#coinbaseAddress in TransactionFragment's

OlivierBBB committed:
fix: use the Hub#coinbaseAddress in TXN_DATA

OlivierBBB committed:
fix: update constraints commit

OlivierBBB committed:
spotless

OlivierBBB committed:
Merge branch 'arith-dev' into 1648-initialization-and-finalization-sections-fix

OlivierBBB committed:
fix: constraints commit update

OlivierBBB committed:
fix(disgusting): coinbaseAddress in TX_INIT + associated reworking

also moving address comparison methods
  from TransactionProcessingMetadata
  to TxSkipSection

OlivierBBB committed:
spotless

DavePearce committed:
update self-hosted runner instance

This updates the self-hosted runner instance for the
`go-corset-replay-tests` job from `large` to `xxl`.  This seems
necessary as the large instance appears to be running out of memory.

DavePearce committed:
update other self-hosted runner instances

This also updates the self-hosted runner instance from large to xxl for
the `corset-unit-tests`, `go-corset-unit-tests` and the
`corset-replay-tests` jobs.  Specifically, the unit tests are taking a
fair amount of time, so this seems sensible.

OlivierBBB committed:
ras: LINEA_GAS_LIMIT to ETHEREUM_GAS_LIMIT

OlivierBBB committed:
ras: back to LINEA_GAS_LIMIT_XXX
